Use regex library and do barcode disambiguation

roryk opened the pull request:
Hi Valentine,

This pull request has two things, you can reject either one or both.

The first is to use the regex library instead of the re library. This allows for syntax to allow mismatches in a regex, so for example:

```
{
    "read1": "(?P<name>[^\\s]+)\\n(?P<CB1>\\w{8,11})(GAGTGATTGCTTGTGACGCCTT){e<=3}(?P<CB2>\\w{8})(?P<MB>\\w{6})(.*)\\n+(.*)\\n(.*)\\n",
    "read2": "(@.*)\\n(?P<seq>.*)\\n\\+(.*)\\n(?P<qual>.*)\\n"
}
```

The `{e<=3}` syntax allows for 3 or less errors.

The second change is to do barcode disambiguation. I abstracted out the writing to either write to barcode files or to stdout depending on if the option is set or not.

roryk committed:
Use the regex library to allow for mismatches.

roryk committed:
Add option to keep dual index barcodes separate during transformation.

roryk commented:
I dropped the barcode disambiguation, it was super slow if there were a lot of barcodes. I added an option to keep dual index barcodes separate. There are some preps that have barcode1 and barcode2 but the combination of barcode1 and barcode2 is random, so enumerating and searching through those possibilities for every read takes too long.

vals commented:
Hi Rory, I just got back to a conference and I'll try to look in to these by Monday or so. (And I'll try push the new version to PyPI)

I didn't know there was an alternative regex library!

My solution for the combined barcodes is to computationally merge combinations in to one. I think I have an example STRT-seq transform that does this.

roryk commented:
Thanks Valentine,

I wanted to allow mismatches in the barcode filtering, but it was super slow when there were tons of barcode combinations with the split barcodes. With the 384 BC1 and 384 BC2 there were 150k combinations to search through and it took forever allowing a mismatch. Keeping them separate would let the lookup happen just through the 384*2.

roryk committed:
Parallelize fastq transformation.

On my laptop with 1 core, 1 million reads:

real	0m37.614s
user	0m35.506s
sys	0m1.313s

2 cores:

real	0m22.627s
user	0m35.284s
sys	0m1.547s

roryk commented:
I added some code to parallelize the FASTQ conversion as well using the multiprocessing library.

roryk committed:
Separate out cb_filter into its own subcommand.

Closes #2.

roryk commented:
I broke out `cb_filter` into its own subcommand.

roryk committed:
Parallelize barcode filtering.

roryk committed:
Use partition_all during transformation.

This avoids dropping incomplete chunks.

roryk commented:
I parallelized the cellular barcode filtering too, so maybe we can put some more expensive mismatch searching in there.

roryk commented:
Sorry for the spammy pull request, I can break these into separate branches if it would be easier to look at/test.

vals commented:
Hi Rory, there were a couple of issues, but now it all works. Just need to remember now how to build for PyPI...
